Add IPv6 support to iptables proxier

leblancd opened the pull request:
Add IPv6 support to iptables proxier

The following changes are proposed for the iptables proxier:

- There are three places where a string specifying IP:port is parsed
  using something like this:
      if index := strings.Index(e.endpoint, ":"); index != -1 {
  This will fail for IPv6 since V6 addresses contain colons. Also,
  the V6 address is expected to be surrounded by square brackets
  (i.e. [<ipv6-addr>]:<port>). Fix this by replacing call to Index with
  call to LastIndex() and stripping out square brackets.
- The String() method for the localPort struct should put square brackets
  around IPv6 addresses.
- The logging in the merge() method for proxyServiceMap should put brackets
  around IPv6 addresses.
- There are several places where filterRules destination is hardcoded to
  <clusterIP>/32. This should be a /128 for IPv6 case.
- Add IPv6 unit test cases

Note: I've left out most of the UT test cases that I had included in my original version of this
PR because the number of lines of code change were much too large for a single review.
I'm including a minimum of UT with this current version of the PR.

fixes #48550

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
This PR addresses several issues in the iptables proxier for handling IPv6 addresses
that were found via visual code inspection, including:
- There are three places where a string specifying IP:port using something like the following:
       if index := strings.Index(e.endpoint, ":"); index != -1 {
  This will fail for IPv6 since V6 addresses contains many colons, and the V6 address is expected
  to be enclosed in square brackets when followed by :<port>.
- The String() method for the localPort struct should put square brackets around IPv6 addresses.
- The logging in the merge() method for proxyServiceMap should put brackets around IPv6
  addresses.
- There are several places where filterRules destination is hardcoded to /32.
  Should be a /128 for IPv6 case.
- More IPv6 unit test cases are needed.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #48550

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

cmluciano commented:
@k8s-bot ok to test

pmichali commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -166,12 +167,26 @@ type endpointsInfo struct {
+	if index := strings.LastIndex(s, ":"); index != -1 {
+		ip := s[0:index]
+		// Strip off any surrounding brackets.
+		re := regexp.MustCompile(`\[(.*)\]`)
```
Would a less compute intensive solution be to just check for a '[' at position zero and then take slice w/o first and last character?

thockin replied:
yes, regex seems overpowered for this

pmichali commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -300,15 +315,26 @@ func (scm *serviceChangeMap) update(namespacedName *types.NamespacedName, previo
+// ipPort returns a string of the form "<ip>:<port>" (for IPv4)
+// or "[<ip>]:<port>" (for IPv6) for a given IP address and port
+func ipPort(ipStr string, port int) string {
+	ip := net.ParseIP(ipStr)
```
Do we know that ipStr will always be an IP address (and not a name)? Is it guaranteed to always be valid?

leblancd replied:
@thockin: Yes, that will do the trick, thanks!

pmichali commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -962,6 +986,16 @@ func (proxier *Proxier) deleteEndpointConnections(connectionMap map[endpointServ
+// IPv4 and <ip-address>/128 for IPv6
+func hostAddress(ip net.IP) string {
+	len := 32
+	if ip.To4() == nil {
```
Is it always guaranteed that ip is a valid IP address? IOW no need to check for nil? Just wondering if all the callers using this are providing valid IP addresses.

leblancd replied:
The nil check isn't being done as a sanity check, it's being done to determine if the IP address is IPv4 or IPv6.

pmichali reviewed (commented):
Bunch of questions in-line. Mostly minor stuff.

pmichali commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1590,30 +1925,53 @@ func makeServiceMap(proxier *Proxier, allServices ...*api.Service) {
 		ept.Subsets = []api.EndpointSubset{{
 			Addresses: []api.EndpointAddress{{
-				IP: "1.1.1.1",
+				IP: testIP(1, 1),
```
I really think the indirection of the testIP function makes the tests harder to read.

pmichali reviewed (changes requested):
Would suggest adding additional test cases that exercise the new/modified methods directly (versus through some call chain of higher level function test). Especially as you can exercise error cases.

leblancd commented:
/retest
(Looks like test failure was flake #48368)

cmluciano commented:
/assign @dcbw

dcbw commented:
Also the pull-kubernetes-verify test failed; that's likely because of new imports to the testcases and proxy code.  A good way to update that is:

`docker run -it -v $(pwd):/go/src/k8s.io/kubernetes -w /go/src/k8s.io/kubernetes golang:1.8 /bin/bash -c "hack/update-bazel.sh"`

And then `git add` any changed files and `git commit --amend` to add them to the existing commit.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -166,12 +167,26 @@ type endpointsInfo struct {
+// Returns just the IP part of an IP:port or IP endpoint string. If the IP
+// part is an IPv6 address enclosed in brackets (e.g. "[fd00:1::5]:9999"),
+// then the brackets are stripped as well.
+func ipPart(s string) string {
```
that seems to DTRT

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -300,15 +315,26 @@ func (scm *serviceChangeMap) update(namespacedName *types.NamespacedName, previo
+// ipPort returns a string of the form "<ip>:<port>" (for IPv4)
+// or "[<ip>]:<port>" (for IPv6) for a given IP address and port
+func ipPort(ipStr string, port int) string {
+	ip := net.ParseIP(ipStr)
```
net.JoinHostPort() seems to do exactly what this does, if we just fmt.Sprintf("%d", port) or stconv.Itoa(port)

thockin reviewed (commented):
Would be nice to see this as individual tiny commits, especially as large as the test changes are.

thockin commented:
ping me when ready for review.

thockin commented:
We have O(days) to get this into v1.8

leblancd commented:
@thockin Thanks. I think this will have to wait until v1.9. I'm seeing some issues in my functional tests with this code. It might be that the code is okay, but I'm driving it incorrectly. Sorry for the delay on this.

thockin commented:
Assign back to me when ready to review.

Commit:
Add IPv6 support to iptables proxier

The following changes are proposed for the iptables proxier:

* There are three places where a string specifying IP:port is parsed
  using something like this:

      if index := strings.Index(e.endpoint, ":"); index != -1 {

  This will fail for IPv6 since V6 addresses contain colons. Also,
  the V6 address is expected to be surrounded by square brackets
  (i.e. []:). Fix this by replacing call to Index with
  call to LastIndex() and stripping out square brackets.
* The String() method for the localPort struct should put square brackets
  around IPv6 addresses.
* The logging in the merge() method for proxyServiceMap should put brackets
  around IPv6 addresses.
* There are several places where filterRules destination is hardcoded to
  /32. This should be a /128 for IPv6 case.
* Add IPv6 unit test cases

fixes #48550

danehans commented:
/area ipv6
/sig network

leblancd commented:
I'll post a rebase of the operational (non-unit-test) code with changes for review comments. For UT, I need to figure out how to split up the test cases into several, smaller-sized reviews (and add minimum set of test cases back into this review).

leblancd commented:
/assign thockin
/unassign leblancd

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -166,12 +166,30 @@ type endpointsInfo struct {
+		// IP address without brackets
+		return s
+	}
+	if s[0] == '[' && s[len(s)-1] == ']' {
```
is this a legit notation?  Why wouldn't `ParseIP` handle it correctly?

leblancd replied:
You're right, this is not legit. The brackets should only appear for IPv6:port or when IPv6 address appears in a URL. I'll delete these lines.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -798,11 +818,13 @@ func getLocalIPs(endpointsMap proxyEndpointsMap) map[types.NamespacedName]sets.S
-				nsn := svcPortName.NamespacedName
-				if localIPs[nsn] == nil {
-					localIPs[nsn] = sets.NewString()
+				if ip := ep.IPPart(); ip != "" {
```
Why would this case happen?  If legit, leave a comment?

leblancd replied:
I suppose this depends on how defensive we want to be with our coding. Here ep.IPPart() could return a null string if net.SplitHostPort() detects bad syntax and returns a null string (along with an error). This can happen (theoretically) if the string in ep.endpoint is not quite right. For example, if the ep.endpoint has too many colons, e.g. "1.2.3.4::9999", or brackets that are out of place, e.g. "1.2.3.4:[0]", then net.SplitHostPort() returns a null string and an error. It's probably very unlikely that we'd end up with a bad format in ep.endpoint at this point, with sanity checks that are done elsewhere. I'll leave this null-string check in for now and add a comment, let me know if you feel this is being too defensive.

thockin reviewed (commented):
Just a couple minor questions, I think.

thockin commented:
/lgtm
/approve
